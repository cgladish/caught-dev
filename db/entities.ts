export type ServiceAuthEntity = {
  id: number;
  appName: AppName;
  encryptedToken: string;
};

export type PreservationRuleEntity = {
  id: number;
  appName: AppName;
  name: string;
  selectedJson: string;
  startDatetime: string | null;
  endDatetime: string | null;
  createdAt: string;
  updatedAt: string;
};

export type MessageEntity = {
  id: number;
  preservationRuleId: number;
  externalId: string;
  externalChannelId: string;
  authorId: string;
  authorName: string;
  authorAvatar?: string;
  content: string;
  sentAt: string;
  appSpecificDataJson?: string;
};
