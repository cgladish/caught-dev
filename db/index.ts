import knex from "knex";
const config = require("./knexfile");

const environment = process.env.JEST_WORKER_ID ? "test" : "production";

const db = knex(config[environment]);
const migrate = db.migrate.latest(config[environment].migrations);

export default async function getDb() {
  await migrate;
  return db;
}
