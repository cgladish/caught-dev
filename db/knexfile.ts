import type { Knex } from "knex";
import path from "path";

const config: Knex.Config = {
  client: "sqlite3",
  connection: {
    filename: path.join(__dirname, "db.sqlite3"),
  },
  migrations: {
    tableName: "migrations",
    directory: path.join(__dirname, "migrations"),
  },
};

module.exports = config;
