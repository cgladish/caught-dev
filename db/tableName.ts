enum TableName {
  ServiceAuth = "ServiceAuth",
  PreservationRule = "PreservationRule",
  Message = "Message",
}

export default TableName;
