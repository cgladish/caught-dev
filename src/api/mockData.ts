import { v4 } from "uuid";
import getDb from "../db";
import {
  ChannelEntity,
  MessageEntity,
  PreservationRuleEntity,
  WordCountEntity,
} from "../db/entities";
import TableName from "../db/tableName";

export const makeMessage = async (
  preservationRuleId: number,
  overwrites?: Partial<MessageEntity>
): Promise<MessageEntity> => {
  const db = await getDb();
  const [id] = await db<MessageEntity>(TableName.Message).insert({
    preservationRuleId,
    externalId: v4(),
    externalChannelId: "1234567",
    authorId: "1234567",
    authorName: "Name",
    content: "Content",
    sentAt: "2022-09-21T02:00:00.000Z",
    appSpecificDataJson: null,
    ...overwrites,
  });
  const entity = await db<MessageEntity>(TableName.Message)
    .where({ id })
    .first();
  return entity!;
};

export const makeWordCount = async (
  preservationRuleId: number,
  overwrites?: Partial<WordCountEntity>
): Promise<WordCountEntity> => {
  const db = await getDb();
  const [id] = await db<WordCountEntity>(TableName.WordCount).insert({
    preservationRuleId,
    word: "word",
    count: 5,
    ...overwrites,
  });
  const entity = await db<WordCountEntity>(TableName.WordCount)
    .where({ id })
    .first();
  return entity!;
};

export const makePreservationRule = async (
  overwrites?: Partial<PreservationRuleEntity>
): Promise<PreservationRuleEntity> => {
  const db = await getDb();
  const [id] = await db<PreservationRuleEntity>(
    TableName.PreservationRule
  ).insert({
    appName: "discord",
    name: v4(),
    selectedJson: '{"key":"value"}',
    startDatetime: "2022-09-21T02:00:00.000Z",
    endDatetime: "2022-09-21T04:00:00.000Z",
    initialBackupComplete: false,
    ...overwrites,
  });
  const entity = await db<PreservationRuleEntity>(TableName.PreservationRule)
    .where({ id })
    .first();
  return entity!;
};

export const makeChannel = async (
  overwrites?: Partial<ChannelEntity>
): Promise<ChannelEntity> => {
  const db = await getDb();
  const [id] = await db<ChannelEntity>(TableName.Channel).insert({
    appName: "discord",
    externalId: v4(),
    name: "channel name",
    iconUrl: null,
    ...overwrites,
  });
  const entity = await db<ChannelEntity>(TableName.Channel)
    .where({ id })
    .first();
  return entity!;
};
