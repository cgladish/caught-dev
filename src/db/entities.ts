export type ServiceAuthEntity = {
  id: number;
  appName: AppName;
  encryptedToken: string;
};

export type PreservationRuleEntity = {
  id: number;
  appName: AppName;
  name: string;
  selectedJson: string;
  startDatetime?: string | null;
  endDatetime?: string | null;
  initialBackupComplete: boolean;
  createdAt: string;
  updatedAt: string;
};

export type ChannelEntity = {
  id: number;
  appName: AppName;
  externalId: string;
  name: string;
  iconUrl?: string | null;
};

export type MessageEntity = {
  id: number;
  preservationRuleId: number;
  externalId: string;
  externalChannelId: string;
  authorId: string;
  authorName: string;
  authorAvatar?: string | null;
  content: string;
  sentAt: string;
  appSpecificDataJson?: string | null;
};

export type WordCountEntity = {
  id: number;
  preservationRuleId: number;
  word: string;
  count: number;
};
