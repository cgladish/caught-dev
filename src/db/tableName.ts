enum TableName {
  ServiceAuth = "ServiceAuth",
  PreservationRule = "PreservationRule",
  Message = "Message",
  MessageSearch = "MessageSearch",
  Channel = "Channel",
  WordCount = "WordCount",
}

export default TableName;
