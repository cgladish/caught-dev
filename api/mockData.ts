import { v4 } from "uuid";
import getDb from "../db";
import { MessageEntity, PreservationRuleEntity } from "../db/entities";
import TableName from "../db/tableName";

export const makeMessage = async (
  preservationRuleId: number,
  overwrites?: Partial<MessageEntity>
): Promise<MessageEntity> => {
  const db = await getDb();
  const [id] = await db<MessageEntity>(TableName.Message).insert({
    preservationRuleId,
    externalId: v4(),
    externalChannelId: "1234567",
    authorId: "1234567",
    authorName: "Name",
    content: "Content",
    sentAt: "2022-09-21T02:00:00.000Z",
    ...overwrites,
  });
  const entity = await db<MessageEntity>(TableName.Message)
    .where({ id })
    .first();
  return entity!;
};

export const makePreservationRule = async (
  overwrites?: Partial<PreservationRuleEntity>
): Promise<PreservationRuleEntity> => {
  const db = await getDb();
  const [id] = await db<PreservationRuleEntity>(
    TableName.PreservationRule
  ).insert({
    appName: "discord",
    name: v4(),
    selectedJson: '{"key":"value"}',
    startDatetime: "2022-09-21T02:00:00.000Z",
    endDatetime: "2022-09-21T04:00:00.000Z",
    initialBackupComplete: false,
    ...overwrites,
  });
  const entity = await db<PreservationRuleEntity>(TableName.PreservationRule)
    .where({ id })
    .first();
  return entity!;
};
