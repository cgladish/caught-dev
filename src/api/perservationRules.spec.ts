import { Knex } from "knex";
import omit from "lodash/omit";
import getDb from "../db";
import { MessageEntity } from "../db/entities";
import TableName from "../db/tableName";
import { makeMessage, makePreservationRule } from "./mockData";
import {
  createPreservationRule,
  updatePreservationRule,
  fetchPreservationRules,
  PreservationRule,
  deletePreservationRule,
} from "./preservationRules";

describe("preservationRules", () => {
  let db: Knex;

  beforeEach(async () => {
    db = await getDb();
  });

  describe("createPreservationRule", () => {
    it("creates a preservation rule", async () => {
      const createdPreservationRule = await createPreservationRule(
        {
          appName: "discord",
          name: "Rule",
          selected: { key: "value" },
          startDatetime: new Date("2022-09-21T02:00:00.000Z"),
          endDatetime: new Date("2022-09-21T04:00:00.000Z"),
        },
        false
      );
      expect(
        omit(createdPreservationRule, "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();

      const preservationRules = await db<PreservationRule>(
        TableName.PreservationRule
      ).select();
      expect(preservationRules.length).toEqual(1);
      expect(
        omit(preservationRules[0], "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();
    });
  });

  describe("updatePreservationRule", () => {
    it("updates a preservation rule", async () => {
      await makePreservationRule({
        name: "Rule",
        startDatetime: "2022-09-21T02:00:00.000Z",
        endDatetime: "2022-09-21T04:00:00.000Z",
      });
      const { id } = await makePreservationRule({
        appName: "twitter" as any,
      });

      const updatedPreservationRule = await updatePreservationRule(id, {
        name: "Rule3",
        selected: { key3: "value3" },
        startDatetime: new Date("2022-09-21T03:00:00.000Z"),
        endDatetime: new Date("2022-09-21T05:00:00.000Z"),
      });
      expect(
        omit(updatedPreservationRule, "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();

      const preservationRules = await db<PreservationRule>(
        TableName.PreservationRule
      ).select();
      expect(preservationRules.length).toEqual(2);
      expect(
        omit(preservationRules[0], "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();
      expect(
        omit(preservationRules[1], "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();
    });
  });

  describe("deletePreservatinoRule", () => {
    it("deletes a preservation rule and associated messages", async () => {
      const { id } = await makePreservationRule({
        appName: "twitter" as any,
      });
      const { id: otherId } = await makePreservationRule({
        name: "Rule",
        startDatetime: "2022-09-21T02:00:00.000Z",
        endDatetime: "2022-09-21T04:00:00.000Z",
      });
      await makeMessage(id);
      await makeMessage(id);
      const { id: otherMessageId } = await makeMessage(otherId);

      await deletePreservationRule(id);

      const preservationRules = await db<PreservationRule>(
        TableName.PreservationRule
      ).select();
      expect(preservationRules.length).toEqual(1);
      expect(preservationRules[0]!.id).toEqual(otherId);

      const messages = await db<MessageEntity>(TableName.Message).select();
      expect(messages.length).toEqual(1);
      expect(messages[0]!.id).toEqual(otherMessageId);
    });
  });

  describe("fetchPreservationRules", () => {
    it("fetches preservation rules", async () => {
      await makePreservationRule({
        name: "Rule",
      });
      await makePreservationRule({
        name: "Rule2",
        selectedJson: '{ "key2": "value2" }',
        startDatetime: "2022-09-21T03:00:00.000Z",
        endDatetime: "2022-09-21T05:00:00.000Z",
      });
      await makePreservationRule({
        appName: "twitter" as any,
      });

      const preservationRules = await fetchPreservationRules("discord");
      expect(preservationRules.length).toEqual(2);
      expect(
        omit(preservationRules[0], "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();
      expect(
        omit(preservationRules[1], "id", "createdAt", "updatedAt")
      ).toMatchSnapshot();
    });
  });
});
