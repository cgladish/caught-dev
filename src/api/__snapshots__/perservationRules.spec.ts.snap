// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preservationRules createPreservationRule creates a preservation rule 1`] = `
{
  "appName": "discord",
  "endDatetime": 2022-09-21T04:00:00.000Z,
  "initialBackupComplete": 0,
  "name": "Rule",
  "selected": {
    "key": "value",
  },
  "startDatetime": 2022-09-21T02:00:00.000Z,
}
`;

exports[`preservationRules createPreservationRule creates a preservation rule 2`] = `
{
  "appName": "discord",
  "endDatetime": "2022-09-21T04:00:00.000Z",
  "initialBackupComplete": 0,
  "name": "Rule",
  "selectedJson": "{"key":"value"}",
  "startDatetime": "2022-09-21T02:00:00.000Z",
}
`;

exports[`preservationRules fetchPreservationRules fetches preservation rules 1`] = `
{
  "appName": "discord",
  "endDatetime": 2022-09-21T04:00:00.000Z,
  "initialBackupComplete": 0,
  "name": "Rule",
  "selected": {
    "key": "value",
  },
  "startDatetime": 2022-09-21T02:00:00.000Z,
}
`;

exports[`preservationRules fetchPreservationRules fetches preservation rules 2`] = `
{
  "appName": "discord",
  "endDatetime": 2022-09-21T05:00:00.000Z,
  "initialBackupComplete": 0,
  "name": "Rule2",
  "selected": {
    "key2": "value2",
  },
  "startDatetime": 2022-09-21T03:00:00.000Z,
}
`;

exports[`preservationRules updatePreservationRule updates a preservation rule 1`] = `
{
  "appName": "twitter",
  "endDatetime": 2022-09-21T05:00:00.000Z,
  "initialBackupComplete": 0,
  "name": "Rule3",
  "selected": {
    "key3": "value3",
  },
  "startDatetime": 2022-09-21T03:00:00.000Z,
}
`;

exports[`preservationRules updatePreservationRule updates a preservation rule 2`] = `
{
  "appName": "discord",
  "endDatetime": "2022-09-21T04:00:00.000Z",
  "initialBackupComplete": 0,
  "name": "Rule",
  "selectedJson": "{"key":"value"}",
  "startDatetime": "2022-09-21T02:00:00.000Z",
}
`;

exports[`preservationRules updatePreservationRule updates a preservation rule 3`] = `
{
  "appName": "twitter",
  "endDatetime": "2022-09-21T05:00:00.000Z",
  "initialBackupComplete": 0,
  "name": "Rule3",
  "selectedJson": "{"key3":"value3"}",
  "startDatetime": "2022-09-21T03:00:00.000Z",
}
`;
